### PR TITLE
Filter OOB points while training

### PR DESCRIPTION
**Pull request by gitttt-1234**
### Description
This PR addresses #1901 and filters all OOb and negative points while generating the training dataset.

### Types of changes

- [ ] Bugfix
- [ ] New feature
- [ ] Refactor / Code style update (no logical changes)
- [ ] Build / CI changes
- [ ] Documentation Update
- [ ] Other (explain)

### Does this address any currently open issues?
#1901 

### Outside contributors checklist

- [ ] Review the [guidelines for contributing](https://github.com/talmolab/sleap/blob/develop/docs/CONTRIBUTING.md) to this repository
- [ ] Read and sign the [CLA](https://github.com/talmolab/sleap/blob/develop/sleap-cla.pdf) and add yourself to the [authors list](https://github.com/talmolab/sleap/blob/develop/AUTHORS)
- [ ] Make sure you are making a pull request against the **develop** branch (not *main*). Also you should start *your branch* off *develop*
- [ ] Add tests that prove your fix is effective or that your feature works
- [ ] Add necessary documentation (if appropriate)

#### Thank you for contributing to SLEAP!
:heart:

### CODE REVIEW

File: sleap/nn/data/providers.py [hunk 1]
```diff
@@ -198,6 +199,25 @@ def py_fetch_lf(ind):
...
+            # Filter OOB points
+            instances = []
+            for instance in lf.instances:
+                pts = instance.numpy()
+                # negative coords
+                pts[pts < 0] = np.NaN
+
+                # coordinates outside img frame
+                pts[:, 0][pts[:, 0] > height - 1] = np.NaN
+                pts[:, 1][pts[:, 1] > width - 1] = np.NaN
```
[the remark this answers is not in the document]
Comment (eberrigan) [reply sleap/nn/data/providers.py:213]: add tests for inside bounds, outside bounds, and edge cases to make sure this behaves as expected

File: sleap/nn/data/providers.py [hunk 2]
```diff
@@ -198,6 +199,25 @@ def py_fetch_lf(ind):
...
+
+                # remove all nans
+                pts = pts[~np.isnan(pts).any(axis=1), :]
```
Comment (eberrigan) [inline sleap/nn/data/providers.py:216]: add tests for when pts becomes all NaNs

File: tests/nn/data/test_providers.py [hunk 3]
```diff
@@ -68,6 +68,37 @@ def test_labels_reader_no_visible_points(min_labels):
...
+    assert all(np.isnan(examples[0]["instances"][0][0]))
+
+    # test with negative keypoints
+    labels = min_labels.copy()
+    assert len(labels.labeled_frames[0].instances) == 2
+
+    assert labels[0].instances[0].numpy().shape[0] == 2  # 2 nodes
+
+    labels[0].instances[0][1] = (-10, 100)
+
```
Comment (eberrigan) [inline tests/nn/data/test_providers.py:94]: use function modified instead of copying code for ease of development in the future

**Review by eberrigan — COMMENTED** [review]
add tests for training with out of bounds points. we can add a fixture for outof bounds points to make testing easier

**Review by eberrigan — APPROVED** [review]
Looks good! Code rabbits suggestions of parameterizing could be implemented.

**Comment by gitttt-1234** [comment]
> Looks good! Code rabbits suggestions of parameterizing could be implemented.

Sure, added those in the latest commit!

File: sleap/nn/data/instance_cropping.py [hunk 4]
```diff
@@ -42,12 +42,23 @@ def find_instance_crop_size(
...
+
+            pts[pts < 0] = np.NaN
+            height, width = lf.image.shape[:2]
+            pts[:, 0][pts[:, 0] > width - 1] = np.NaN
+            pts[:, 1][pts[:, 1] > height - 1] = np.NaN
+
+            pts *= input_scaling
+            max_length = np.maximum(
+                max_length, np.nanmax(pts[:, 0]) - np.nanmin(pts[:, 0])
+            )
```
Comment (roomrys) [inline sleap/nn/data/instance_cropping.py]: This leaves a possibility (albeit unlikely) for a `max_length` of `np.nan` if we have an all-nan instance for some or other reason (possibly because of the rumours spread [here](https://github.com/talmolab/sleap/issues/640#issuecomment-1419990083)) - although, on the same point, we currently just filter nan instances. 

Even so, suggestion:
```suggestion
            max_length: float = np.nanmax(
                [max_length, np.nanmax(pts[:, 0]) - np.nanmin(pts[:, 0])]
            )
```

File: sleap/nn/data/instance_cropping.py [hunk 5]
```diff
@@ -42,12 +42,23 @@ def find_instance_crop_size(
...
     min_crop_size_no_pad = min_crop_size - padding
     max_length = 0.0
-    for inst in labels.user_instances:
-        pts = inst.points_array
-        pts *= input_scaling
-        max_length = np.maximum(max_length, np.nanmax(pts[:, 0]) - np.nanmin(pts[:, 0]))
-        max_length = np.maximum(max_length, np.nanmax(pts[:, 1]) - np.nanmin(pts[:, 1]))
-        max_length = np.maximum(max_length, min_crop_size_no_pad)
+    for lf in labels:
+        for inst in lf.user_instances:
```
Comment (roomrys) [inline sleap/nn/data/instance_cropping.py]: It would be faster to
```suggestion
    for lf in labels:
        for inst in lf:
            if isinstance(inst, sleap.PredictedInstance):
                continue
```
since
https://github.com/talmolab/sleap/blob/66d96ced50153e98b952176fbdb73043e105e397/sleap/instance.py#L1552-L1555

File: sleap/nn/data/providers.py [hunk 6]
```diff
@@ -198,34 +199,61 @@ def py_fetch_lf(ind):
...
+            for inst in insts:
+
+                # Filter OOB
+                pts = inst.numpy()
+                pts[pts < 0] = np.NaN
+
+                pts[:, 0][pts[:, 0] > width - 1] = np.NaN
+                pts[:, 1][pts[:, 1] > height - 1] = np.NaN
+
+                instance = Instance.from_numpy(pts, inst.skeleton, inst.track)
```
Comment (roomrys) [inline sleap/nn/data/providers.py:215]: Upon further investigation, `Instance.numpy` returns a copy of the underlying `Instance._points`, so no need to create a new `Instance.from_numpy`
```suggestion
```

File: sleap/nn/data/providers.py [hunk 7]
```diff
@@ -198,34 +199,61 @@ def py_fetch_lf(ind):
...
+            instances = []
+
+            for inst in insts:
+
+                # Filter OOB
+                pts = inst.numpy()
+                pts[pts < 0] = np.NaN
+
+                pts[:, 0][pts[:, 0] > width - 1] = np.NaN
+                pts[:, 1][pts[:, 1] > height - 1] = np.NaN
```
Comment (roomrys) [inline sleap/nn/data/providers.py]: If we are doing this in multiple places, it would be nice to have it as a re-useable function.

**Review by roomrys — CHANGES_REQUESTED** [review]
